File: android/service/platform_service.cpp
/*
 * Copyright (C) 2016 Simon Fels <morphis@gravedo.de>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3, as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranties of
 * MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "android/service/platform_service.h"
#include "anbox/rpc/channel.h"

#include "anbox_rpc.pb.h"
#include "anbox_bridge.pb.h"

#define LOG_TAG "Anboxd"
#include <cutils/log.h>

#include <utils/String8.h>

using namespace android;

namespace android {
PlatformService::PlatformService(const std::shared_ptr<anbox::PlatformApiStub> &platform_api_stub) :
    platform_api_stub_(platform_api_stub) {
}

status_t PlatformService::boot_finished() {
    platform_api_stub_->boot_finished();
    return OK;
}

anbox::PlatformApiStub::WindowStateUpdate::Window PlatformService::unpack_window_state(const Parcel &data) {
    bool has_surface = data.readByte() != 0;

    String8 package_name(data.readString16());

    auto frame_left = data.readInt32();
    auto frame_top = data.readInt32();
    auto frame_right = data.readInt32();
    auto frame_bottom = data.readInt32();
    auto task_id = data.readInt32();
    auto stack_id = data.readInt32();
    auto rotation_angle = data.readInt32();

    return anbox::PlatformApiStub::WindowStateUpdate::Window{
        -1, // Display id will be added by the caller
        has_surface,
        package_name.string(),
        {frame_left, frame_top, frame_right, frame_bottom},
        task_id,
        stack_id,
    };
}

status_t PlatformService::update_window_state(const Parcel &data) {
    anbox::PlatformApiStub::WindowStateUpdate state;

    const auto num_displays = data.readInt32();
    for (auto n = 0; n < num_displays; n++) {
        const auto display_id = data.readInt32();
        const auto num_windows = data.readInt32();
        for (auto m = 0; m < num_windows; m++) {
            auto window = unpack_window_state(data);
            window.display_id = display_id;
            state.updated_windows.push_back(window);
        }
    }

    const auto num_removed_windows = data.readInt32();
    for (auto n = 0; n < num_removed_windows; n++) {
        auto window = unpack_window_state(data);
        state.removed_windows.push_back(window);
    }

    platform_api_stub_->update_window_state(state);

    return OK;
}

status_t PlatformService::update_application_list(const Parcel &data) {
    anbox::PlatformApiStub::ApplicationListUpdate update;
    const auto num_packages = data.readInt32();
    for (auto n = 0; n < num_packages; n++) {
        String8 name(data.readString16());
        String8 package_name(data.readString16());

        auto p = anbox::PlatformApiStub::ApplicationListUpdate::Application{
                name.string(),
                package_name.string(),
        };

        String8 action(data.readString16());
        String8 uri(data.readString16());
        String8 type(data.readString16());
        String8 component_package(data.readString16());
        String8 component_class(data.readString16());

        std::vector<std::string> categories;
        unsigned int num_categories = data.readInt32();
        for (int m = 0; m < num_categories; m++)
            categories.push_back(String8(data.readString16()).string());

        p.launch_intent.action = action;
        p.launch_intent.uri = uri;
        p.launch_intent.type = type;
        p.launch_intent.package = component_package;
        p.launch_intent.component = component_class;
        p.launch_intent.categories = categories;

        data.readByteVector(&p.icon);

        // Send updates with icons separately to not overflow protobuf
        if (p.icon.size() > 0) {
          anbox::PlatformApiStub::ApplicationListUpdate with_icon_update;
          with_icon_update.applications.push_back(p);
          platform_api_stub_->update_application_list(with_icon_update);
        } else {
          update.applications.push_back(p);
        }
    }

    const auto num_removed_packages = data.readInt32();
    for (auto n = 0; n < num_removed_packages; n++) {
      String8 package_name(data.readString16());
      update.removed_applications.push_back(package_name.string());
    }

    platform_api_stub_->update_application_list(update);

    return OK;
}

status_t PlatformService::set_clipboard_data(const Parcel &data) {
    if (data.readInt32() == 0)
      return OK;
    String8 text(data.readString16());
    const auto clip_data = anbox::PlatformApiStub::ClipboardData{text.string()};
    platform_api_stub_->set_clipboard_data(clip_data);
    return OK;
}

status_t PlatformService::get_clipboard_data(const Parcel &data, Parcel *reply) {
    (void) data;
    const auto clip_data = platform_api_stub_->get_clipboard_data();
    if (clip_data.text.empty()) {
      reply->writeInt32(0);
      return OK;
    }
    reply->writeInt32(1);
    reply->writeString16(String16(clip_data.text.c_str(), clip_data.text.length()));
    return OK;
}
} // namespace android

File: android/appmgr/src/org/anbox/appmgr/PlatformService.java
/*
 * Copyright (C) 2016 Simon Fels <morphis@gravedo.de>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3, as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranties of
 * MERCHANTABILITY, SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.anbox.appmgr;

import android.os.ServiceManager;
import android.os.IBinder;
import android.os.Parcel;
import android.os.RemoteException;
import android.util.Log;
import android.content.Intent;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.ApplicationInfo;
import android.net.Uri;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.Canvas;

import java.util.List;
import java.io.ByteArrayOutputStream;

public final class PlatformService {
    private static final String TAG = "AnboxAppMgr";
    private static final String SERVICE_NAME = "org.anbox.PlatformService";
    private static final String DESCRIPTOR = "org.anbox.IPlatformService";

    private static final int TRANSACTION_updateApplicationList = (IBinder.FIRST_CALL_TRANSACTION + 2);

    private IBinder mService = null;
    private PackageManager mPm = null;

    private void connectService() {
        if (mService != null)
            return;
        mService = ServiceManager.getService(SERVICE_NAME);
    }

    public PlatformService(Context context) {
        if (context != null) {
            mPm = context.getPackageManager();
        } else {
            Log.w(TAG, "No context available");
        }

        Log.i(TAG, "Connected to platform service");
    }

    public void notifyPackageRemoved(String packageName) {
        connectService();

        if (mService == null)
            return;

        Log.i(TAG, "Sending package removed notification to host service");

        Parcel data = Parcel.obtain();
        data.writeInterfaceToken(DESCRIPTOR);
        // No added or updated applications to report
        data.writeInt(0);
        // .. but a single removed application
        data.writeInt(1);
        data.writeString(packageName);

        Parcel reply = Parcel.obtain();
        try {
            mService.transact(TRANSACTION_updateApplicationList, data, reply, 0);
        }
        catch (RemoteException ex) {
            Log.w(TAG, "Failed to send updatePackageList request to remote binder service: " + ex.getMessage());
        }
    }

    public void sendApplicationListUpdate() {
        connectService();

        if (mPm == null || mService == null)
            return;

        Parcel data = Parcel.obtain();
        data.writeInterfaceToken(DESCRIPTOR);

        List<ApplicationInfo> apps = mPm.getInstalledApplications(0);
        data.writeInt(apps.size());
        for (int n = 0; n < apps.size(); n++) {
            ApplicationInfo appInfo = apps.get(n);

            Intent launchIntent = mPm.getLaunchIntentForPackage(appInfo.packageName);
            if (launchIntent == null)
                continue;

            Drawable icon = null;
            try {
                icon = mPm.getApplicationIcon(appInfo.packageName);
            }
            catch (PackageManager.NameNotFoundException ex) {
                continue;
            }

            if (icon == null)
                continue;

            String name = appInfo.name;
            CharSequence label = appInfo.loadLabel(mPm);
            if (label != null)
                name = label.toString();

            data.writeString(name);
            data.writeString(appInfo.packageName);

            data.writeString(launchIntent.getAction());
            if (launchIntent.getData() != null)
                data.writeString(launchIntent.getData().toString());
            else
                data.writeString("");
            data.writeString(launchIntent.getType());
            data.writeString(launchIntent.getComponent().getPackageName());
            data.writeString(launchIntent.getComponent().getClassName());
            data.writeInt(launchIntent.getCategories().size());
            for (String category : launchIntent.getCategories())
                data.writeString(category);

            Bitmap iconBitmap = drawableToBitmap(icon);
            ByteArrayOutputStream outStream = new ByteArrayOutputStream();
            iconBitmap.compress(Bitmap.CompressFormat.PNG, 90, outStream);
            data.writeByteArray(outStream.toByteArray());
        }

        // We don't have any removed applications to include in the update
        data.writeInt(0);

        Parcel reply = Parcel.obtain();
        try {
            mService.transact(TRANSACTION_updateApplicationList, data, reply, 0);
        }
        catch (RemoteException ex) {
            Log.w(TAG, "Failed to send updatePackageList request to remote binder service: " + ex.getMessage());
        }
    }

    private Bitmap drawableToBitmap(Drawable drawable) {
        if (drawable instanceof BitmapDrawable)
            return ((BitmapDrawable)drawable).getBitmap();

        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight(), Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);

        return bitmap;
    }
}
